### <font color = green> **ANALYSIS OF O-LEVEL MATHEMATICS EXAMS SCORE OF SECONDARY 4/5 STUDENTS**</br></font>

##### <font color='blue'>**IMPORTANT INFORMATION ABOUT O-LEVEL AND SECONDARY SCHOOL EDUCATION IN SINGAPORE**</font><br>

**CONTEXT:** </br>
- In Singapore, "O-level" stands for the Singapore-Cambridge General Certificate of Education Ordinary Level. </br>
- It is a major national exam taken by students at the end of Secondary 4 or Secondary 5. </br>
- The results decide which schools or programs students can join next, such as Junior Colleges or Polytechnics. </br>

**KEY DETAILS:** </br>
- Age and Timing: Students usually take the exams at age 16 in October and November. </br>
- Secondary 4 / 5: Typically 15 to 17 years old depending on whether they are on the 4-year Express track or 5-year Normal Academic/Technical track. </br>
- Who runs it: Managed by the Ministry of Education (MOE), the Singapore Examinations and Assessment Board (SEAB), and Cambridge International Education. </br>
- Subjects: Students take 5 to 9 subjects, including English, Mother Tongue, Math, Science, and Humanities. </br>

**GRADING AND SCORING:** </br>
- Grades: Scored from A1 (best) to F9 (fail). A grade of C6 or better is a pass. </br>
- Aggregates: Students use scores like L1R5 (Language 1 + 5 Relevant subjects) or L1R4 to apply for post-secondary education. Lower numbers are better. </br>

**RAW EXAM MARKS OR PERCENTAGES:** </br>
- If the range represents the actual marks scored out of 100 on a school mock exam or the estimated raw percentage on the national GCE O-Level Mathematics papers: </br>
- 32 out of 100: This represents a failing score (typically an E8 or F9 grade). It is completely plausible for a student who is struggling significantly with the syllabus or who leaves major sections of the paper blank. </br>
- 100 out of 100: This represents a perfect score. In O-Level Mathematics (especially Elementary Mathematics), achieving full marks or very close to it is common among top-performing students due to the precise, objective nature of the marking scheme. </br>

**CLASS AND COHORT BREAKDOWN:** </br>
- Average Class Size: ~33 students per form class across secondary levels. </br>
- Secondary 4 Cohort: ~39,600 students nationally (spanning Express, Normal (Academic), and Normal (Technical) streams). </br>
- Secondary 5 Cohort: ~2,000 students nationally (restricted to Normal (Academic) students who progress past their N-Levels). </br>

**TYPICAL DAILY SCHEDULE STRUCTURE:** </br>
- Morning Assembly: 7:30 AM – 7:45 AM (Flag raising and attendance taking). </br>
- Recess / Lunch Breaks: 30 to 40 minutes total, usually spaced out mid-morning. </br>
- Standard Dismissal: 2:15 PM – 3:30 PM on most days. 
- Upper secondary students taking extra subjects (e.g., Higher Mother Tongue, pure sciences, or coursework) often stay until 4:00 PM or later on certain weekdays. </br>

**EXTENDED HOURS & AFTER-SCHOOL ACTIVITIES:** </br>
- The academic day frequently extends past standard classroom hours for graduating students: </br>
- Co-Curricular Activities (CCA): Occur 1 to 2 days a week, typically running from 3:30 PM to 6:00 PM.  </br>
- Note that many Sec 4/5 students stand down from CCAs by April or May to focus entirely on their examinations. </br>
- Remedial & Extension Classes: Graduating classes regularly attend structured academic support or preparatory mock exams after regular hours, lasting until 4:30 PM or 5:00 PM </br>

**INDEPENDENT STUDY & TUITION:** </br>
- A 0-20 study hours range referring to homework, tuition, and exam revision outside of school hours is a very accurate reflection of a graduating student's life. </br>
- 0 hours: This only happens during rare school holiday breaks or immediately after major exams. </br>
- 10 to 20 hours: This is the standard reality for most upper secondary students.  </br>
- Spreading 15 hours across a week equates to roughly 2 hours on weekdays and 2.5 hours a day on weekends, which is typical for O-Level or SEC preparation.  </br>

**ATTENDANCE RATE:** </br>
- For a typical student in Singapore, the actual realistic attendance range is tightly bound: </br>
- 95% – 100% (Excellent): The norm for the vast majority of graduating students. </br>
- 90% – 95% (Acceptable): Generally permissible if the minor absences are covered by official Medical Certificates (MCs) or valid letters from parents for compassionate leave. </br>
- Below 90% (Flagged): Any drop below this line immediately triggers an internal school review, requiring formal documentation and meetings with parents. </br>

**Why 40% Attendance is Implausible?**
- Strict MOE Tracking: The Ministry of Education (MOE) monitors school attendance meticulously. Statistically, only about 2% of secondary school students miss more than 10% of school days (falling below 90% attendance) without a valid medical or personal reason.
- Severe Truancy Interventions: A student dropping to a 40% attendance rate implies they are missing roughly three out of every five school days. This status activates the school's welfare framework, involving school counselors, disciplinary actions, home visits, and multi-agency support via programs like the MOE UPLIFT Initiative.
- National Examination Eligibility: Secondary 4 and 5 are graduating years. Mainstream schools and national examination boards generally require high attendance to certify that a student has completed the academic syllabus. Sustained unexcused absences can result in a student being barred from sitting for major national exams like the GCE O-Levels or SEC, or being retained to repeat the academic year.

### <font color = green> **PART A: DATA PREPARATION & CLEAN-UP**</br></font>

In [ ]:
import sqlite3
import pandas as pd
import numpy as np

##### **STEP 1. Connect to the database and explore table(s)**

In [57]:
conn = sqlite3.connect("data/score.db")

tables = pd.read_sql_query(
    "SELECT name FROM sqlite_master WHERE type='table';",
    conn
)

display(tables)

,name
0,score


##### **STEP 2. Load the score table to a Pandas DataFrame**

In [58]:
# df_raw = original data from the source
# df = working copy for EDA 

df_raw = pd.read_sql_query("SELECT * FROM score", conn)

df = df_raw.copy()

df.head(10)

,index,number_of_siblings,direct_admission,CCA,learning_style,student_id,gender,tuition,final_test,n_male,n_female,age,hours_per_week,attendance_rate,sleep_time,wake_time,mode_of_transport,bag_color
0,0,0,Yes,Sports,Visual,ACN2BE,Female,No,69.0,14.0,2.0,16.0,10.0,91.0,22:00,6:00,private transport,yellow
1,1,2,No,Sports,Auditory,FGXIIZ,Female,No,47.0,4.0,19.0,16.0,7.0,94.0,22:30,6:30,private transport,green
2,2,0,Yes,None,Visual,B9AI9F,Male,No,85.0,14.0,2.0,15.0,8.0,92.0,22:30,6:30,private transport,white
3,3,1,No,Clubs,Auditory,FEVM1T,Female,Yes,64.0,2.0,20.0,15.0,18.0,NaN,21:00,5:00,public transport,yellow
4,4,0,No,Sports,Auditory,AXZN2E,Male,No,66.0,24.0,3.0,16.0,7.0,95.0,21:30,5:30,public transport,yellow
5,5,0,No,Arts,Visual,BA6R14,Female,No,57.0,9.0,12.0,15.0,11.0,96.0,22:30,6:30,private transport,red
6,6,2,Yes,None,Visual,D5WGTI,Male,No,69.0,12.0,3.0,16.0,15.0,93.0,21:30,5:30,public transport,green
7,7,0,No,Sports,Visual,HTP8CW,Male,No,76.0,20.0,2.0,15.0,3.0,97.0,21:00,5:00,public transport,green
8,8,0,No,Arts,Auditory,U3YRTC,Male,No,57.0,20.0,7.0,15.0,15.0,98.0,22:00,6:00,private transport,red
9,9,2,No,Arts,Auditory,3MOMA6,Male,Yes,60.0,13.0,9.0,16.0,16.0,NaN,22:30,6:30,private transport,green


##### **STEP 3. Perform Exploratory Data Analysis on the score table**

##### A. Check number of rows and columns

In [ ]:
print(f"Number of rows: {df.shape[0]}")
print(f"Number of columns: {df.shape[1]}")

Number of rows: 15900
Number of columns: 18


##### B. Check column names, dataypes, non-null counts

In [ ]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 15900 entries, 0 to 15899
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   index               15900 non-null  int64  
 1   number_of_siblings  15900 non-null  int64  
 2   direct_admission    15900 non-null  str    
 3   CCA                 15900 non-null  str    
 4   learning_style      15900 non-null  str    
 5   student_id          15900 non-null  str    
 6   gender              15900 non-null  str    
 7   tuition             15900 non-null  str    
 8   final_test          15405 non-null  float64
 9   n_male              15900 non-null  float64
 10  n_female            15900 non-null  float64
 11  age                 15900 non-null  float64
 12  hours_per_week      15900 non-null  float64
 13  attendance_rate     15122 non-null  float64
 14  sleep_time          15900 non-null  str    
 15  wake_time           15900 non-null  str    
 16  mode_of_transpo

##### C. Check for missing values

In [ ]:
df.isnull().sum()

index                   0
number_of_siblings      0
direct_admission        0
CCA                     0
learning_style          0
student_id              0
gender                  0
tuition                 0
final_test            495
n_male                  0
n_female                0
age                     0
hours_per_week          0
attendance_rate       778
sleep_time              0
wake_time               0
mode_of_transport       0
bag_color               0
dtype: int64

##### C. Check for duplicate records

In [26]:
# Check for duplicate records
print("Duplicate rows:", df.duplicated().sum())

# Check for duplicate index
print("Duplicate index:", df["index"].duplicated().sum())

# Check for duplicate student IDs
print("Duplicate student IDs:", df["student_id"].duplicated().sum())

Duplicate rows: 0
Duplicate index: 0
Duplicate student IDs: 900


##### D. View summary statistics (For Numerical columns)

In [ ]:
df.describe()

,index,number_of_siblings,final_test,n_male,n_female,age,hours_per_week,attendance_rate
count,15900.000000,15900.000000,15405.000000,15900.000000,15900.000000,15900.000000,15900.000000,15122.000000
mean,7949.500000,0.886541,67.165401,13.880000,8.906038,15.213459,10.312579,93.270268
std,4590.078975,0.751346,13.977879,6.552584,6.663852,1.758941,4.461861,7.984230
min,0.000000,0.000000,32.000000,0.000000,0.000000,-5.000000,0.000000,40.000000
25%,3974.750000,0.000000,56.000000,10.000000,4.000000,15.000000,7.000000,92.000000
50%,7949.500000,1.000000,68.000000,14.000000,8.000000,15.000000,9.000000,95.000000
75%,11924.250000,1.000000,78.000000,18.000000,13.000000,16.000000,14.000000,97.000000
max,15899.000000,2.000000,100.000000,31.000000,31.000000,16.000000,20.000000,100.000000


##### E. View the values of the Categorical columns

In [33]:
categorical_columns = df.select_dtypes(include="object").columns

for column in categorical_columns:
    print("=" * 50)
    print(f"Column: {column}")
    print("=" * 50)

    # Value counts
    print(df[column].value_counts(dropna=False))

    # % Breakdown
    print("\nPercentage Breakdown:")
    print(round(df[column].value_counts(normalize=True, dropna=False) * 100, 2))

    # # of Unique Values
    print("\nNumber of Unique Values:")
    print(df[column].nunique())

    print("\n")

Column: direct_admission
direct_admission
No     11195
Yes     4705
Name: count, dtype: int64

Percentage Breakdown:
direct_admission
No     70.41
Yes    29.59
Name: proportion, dtype: float64

Number of Unique Values:
2


Column: CCA
CCA
Clubs     3912
Sports    3865
None      3829
Arts      3785
CLUBS      143
NONE       130
ARTS       128
SPORTS     108
Name: count, dtype: int64

Percentage Breakdown:
CCA
Clubs     24.60
Sports    24.31
None      24.08
Arts      23.81
CLUBS      0.90
NONE       0.82
ARTS       0.81
SPORTS     0.68
Name: proportion, dtype: float64

Number of Unique Values:
8


Column: learning_style
learning_style
Auditory    9132
Visual      6768
Name: count, dtype: int64

Percentage Breakdown:
learning_style
Auditory    57.43
Visual      42.57
Name: proportion, dtype: float64

Number of Unique Values:
2


Column: student_id
student_id
AXZN2E    2
T4U89N    2
G0JBVS    2
HJ6Y5P    2
PQ059G    2
         ..
XPECN2    1
7AMC7S    1
XKZ6VN    1
2OU4UQ    1
D9OKLV    1


/var/folders/5x/h0y6d2h9185gj8rngbqzdl4r0000gn/T/ipykernel_2990/900568097.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_columns = df.select_dtypes(include="object").columns


##### <font color='blue'>**SUMMARY OF DATASET CHARACTERISTICS:**</font><br>

Number of rows: 15,900 </br>
- 0 duplicate rows
- index: unique running number

Number of columns: 18 </br>
- 8 numerical
- 10 string (categorical)

##### <font color='blue'>**KEY OBSERVATIONS FOR NUMERICAL COLUMNS:**</font><br>

**DATA ISSUES:** </br>
1. final_test: Target variable has 495 missing records (3.11%)
2. attendance_rate: 778 missing records (4.9%)
3. age: minimum has a value of -5 --Invalid data: For further investigation
4. age range: -5 to 16 years old

**DATA RANGES & OTHER OBSERVATIONS:**
1. index: 0 to 16899 and no duplicate records >> Unique running number
1. number_of_siblings: 0 to 2 >> Plausible
2. final_test range: 32 to 100 >> Plausible
3. n_male: 0 to 31 >> Plausible
4. n_female: 0 to 31 >> Plausible
5. age: mean is 15 >> Plausible
6. hours_per_week range: 0 to 20 study hours >> Plausible
7. attendance_rate range: 40% to 100% >> Plausible range is 90%-100% <font color='red'>KIV how attendance rate below 90% affects final_test</font>

##### <font color='blue'>**KEY OBSERVATIONS FOR TEXT COLUMNS:**</font><br>

**DATA ISSUES:** </br>
1. student ID: 900 duplicate records (5.66%) --For further investigation

**TO BE STANDARDISED:**
1. CCA
2. tuition

**TO BE FEATURE-ENGINEERED:**
1. sleep_time
2. wake_time

**No further data clean-up or transformation to be done**
1. direct_admission: Yes / No
2. learning_style: Auditory / Visual
3. gender: Male / Female
4. mode_of_transport: public transport / private transport / walk
5. bag_color: yellow / green / black / blue / red / white


##### **Step 4. Investigate columns with data issues**
Workflow: Detect → Investigate → Decide → Document → Transform </br>

**<font color='red'>1. DUPLICATE STUDENT IDS**</font> </br>

Number of rows: 900 </br>
% of data: 5.66% </br>

In [ ]:
# Number of unique students
print("Total # of Rows:", len(df))
print("Unique student IDs:", df["student_id"].nunique())

# Show all records belonging to duplicated student IDs
duplicate_students = df[
    df["student_id"].duplicated(keep=False)
].sort_values("student_id")

duplicate_students.head(20)

Total # of Rows: 15900
Unique student IDs: 15000


,index,number_of_siblings,direct_admission,CCA,learning_style,student_id,gender,tuition,final_test,n_male,n_female,age,hours_per_week,attendance_rate,sleep_time,wake_time,mode_of_transport,bag_color
5534,5534,0,No,Clubs,Auditory,00811H,Female,Yes,88.0,21.0,4.0,15.0,8.0,92.0,23:00,7:00,walk,green
12290,12290,0,No,Clubs,Auditory,00811H,Female,Yes,88.0,21.0,4.0,15.0,8.0,92.0,23:00,7:00,walk,white
12270,12270,1,No,Arts,Visual,0195IO,Female,No,52.0,8.0,22.0,16.0,15.0,99.0,22:00,6:00,private transport,yellow
13541,13541,1,No,Arts,Visual,0195IO,Female,No,52.0,8.0,22.0,16.0,15.0,99.0,22:00,6:00,private transport,yellow
4303,4303,0,No,Clubs,Auditory,02RSAH,Female,Yes,64.0,12.0,9.0,15.0,17.0,97.0,22:00,6:00,private transport,yellow
9172,9172,0,No,Clubs,Auditory,02RSAH,Female,Yes,64.0,12.0,9.0,15.0,17.0,97.0,22:00,6:00,private transport,green
8641,8641,0,No,Arts,Auditory,0FC4RY,Female,No,68.0,23.0,4.0,15.0,6.0,NaN,23:00,7:00,walk,blue
9791,9791,0,No,Arts,Auditory,0FC4RY,Female,No,68.0,23.0,4.0,15.0,6.0,100.0,23:00,7:00,walk,yellow
1405,1405,0,Yes,None,Visual,0JUGAF,Female,No,84.0,18.0,2.0,16.0,10.0,94.0,23:00,7:00,walk,red
6339,6339,0,Yes,None,Visual,0JUGAF,Female,No,84.0,18.0,2.0,16.0,10.0,94.0,23:00,7:00,walk,yellow


In [ ]:
# Count how many student IDs are appearing once, twice and more than 2x
id_counts = df["student_id"].value_counts()

print("IDs appearing once:", (id_counts == 1).sum())
print("IDs appearing twice:", (id_counts == 2).sum())
print("IDs appearing >2 times:", (id_counts > 2).sum())

IDs appearing once: 14100
IDs appearing twice: 900
IDs appearing >2 times: 0


In [ ]:
# Identify which columns differ within duplicated student IDs

comparison = (
    df[df["student_id"].duplicated(keep=False)]
    .groupby("student_id")
    .nunique(dropna=False)
)

difference_counts = (
    (comparison.drop(columns="index") > 1)
    .sum()
    .sort_values(ascending=False)
)

display(difference_counts)

bag_color             729
attendance_rate        90
final_test             52
number_of_siblings      0
direct_admission        0
CCA                     0
learning_style          0
gender                  0
tuition                 0
n_male                  0
n_female                0
age                     0
hours_per_week          0
sleep_time              0
wake_time               0
mode_of_transport       0
dtype: int64

In [ ]:
# Count the number of records with duplicated student IDs that have different values in other fields

conflicting_ids = comparison.drop(columns="index").index[
    (comparison.drop(columns="index") > 1).any(axis=1)
]

print(
    "Duplicated students with conflicting data:",
    len(conflicting_ids)
)

Duplicated students with conflicting data: 761


##### <font color='blue'>**KEY OBSERVATIONS for Duplicated records of Student ID:**</font><br>

1. bag_color: 729 records | 80% of the duplicated records >> For further investigation
2. attendance_rate: 90 records >> For further investigation
3. final_test: 52 records >> For further investigation
4. Remaining fields: Identical within duplicate pairs >> No further action required

**<font color='red'>a. Investigate the duplicate student ID records with different bag_color values (729 records)**</font> </br>

In [ ]:
# View the data of the duplicate student ID records with different bag_color values
attendance_conflicts = df[
    df["student_id"].isin(
        comparison.index[comparison["bag_color"] > 1]
    )
].sort_values("student_id")

attendance_conflicts[
    ["student_id", "bag_color"]
].head(20)

,student_id,bag_color
5534,00811H,green
12290,00811H,white
4303,02RSAH,yellow
9172,02RSAH,green
8641,0FC4RY,blue
9791,0FC4RY,yellow
1405,0JUGAF,red
6339,0JUGAF,yellow
9156,0K5LDV,white
15721,0K5LDV,red


In [ ]:
# View bag_color values per student_id

attendance_summary = (
    attendance_conflicts
    .groupby("student_id")["bag_color"]
    .agg(list)
)

attendance_summary.to_csv('duplicates_bag_color.csv')

In [ ]:
# Check if exam scores differ materially across bag colours:
bag_analysis = (
    df.groupby("bag_color")["final_test"]
      .agg(["count", "mean", "median", "std"])
      .round(2)
      .sort_values("mean", ascending=False)
)

bag_analysis

,count,mean,median,std
bag_color,,,,
red,2541,67.43,68.0,13.96
yellow,2658,67.37,68.0,14.14
black,2570,67.36,68.0,13.87
white,2531,67.14,68.0,13.91
blue,2550,67.11,68.0,13.83
green,2555,66.56,67.0,14.13


In [ ]:
# Calculate the range of group means:

mean_range = (
    bag_analysis["mean"].max()
    - bag_analysis["mean"].min()
)

print("Difference between highest and lowest mean:", mean_range)

Difference between highest and lowest mean: 0.8700000000000045


##### <font color='blue'>**KEY OBSERVATIONS for bag_color:**</font><br>

1. Data reliability: Among the 900 duplicated student IDs, 729 (81%) have inconsistent bag_color values. Therefore, it is not a stable characteristic in this dataset.
2. Empirical relevance: Average final-test scores vary by only 0.87 points across the 6 bag colours, against within-colour standard deviations of approximately 14 points.
3. Domain relevance: bag_color showed limited analytical relevance to mathematics performance.
4. Decision: To exclude this feature from predictive modelling to reduce noise and the risk of spurious relationships.

Spurious Relationship: An apparent connection between two variables that looks meaningful in the data but is actually coincidental, caused by another hidden factor, or unlikely to hold in new data. </br>
The trap with bag_color is thinking that any difference in average scores means the variable has predictive meaning; your results show small differences, but that does not mean bag colour has anything to do with mathematics ability. </br>

For example, Students with red bags scored 0.87 points higher on average than students with green bags. </br>

It would be unreasonable to conclude that “Carrying a red bag improves mathematics performance.” </br>
That would confuse association with causation. </br>

**<font color='red'>b. Investigate the duplicate student ID records with different attendance_rate values (90 records)**</font> </br>

In [ ]:
# View the data of the duplicate student ID records with different attendance_rate values

attendance_conflicts = df[
    df["student_id"].isin(
        comparison.index[comparison["attendance_rate"] > 1]
    )
].sort_values("student_id")

attendance_conflicts[
    ["student_id", "attendance_rate"]
].head(20)

,student_id,attendance_rate
8641,0FC4RY,NaN
9791,0FC4RY,100.0
11598,0VTYP7,99.0
12166,0VTYP7,NaN
8240,11RQTT,NaN
9683,11RQTT,98.0
5988,1ICS4I,NaN
13626,1ICS4I,95.0
6062,1KGNR8,88.0
15547,1KGNR8,NaN


In [ ]:
# View attendance_rate values per student_id

attendance_summary = (
    attendance_conflicts
    .groupby("student_id")["attendance_rate"]
    .agg(list)
)

attendance_summary.head(90)

student_id
0FC4RY    [nan, 100.0]
0VTYP7     [99.0, nan]
11RQTT     [nan, 98.0]
1ICS4I     [nan, 95.0]
1KGNR8     [88.0, nan]
262L6N     [nan, 99.0]
2HD89P     [91.0, nan]
2KKLZW     [nan, 99.0]
34XQIE     [nan, 91.0]
35QTZ0     [nan, 94.0]
3HO6WS     [98.0, nan]
3UOMLE     [nan, 70.0]
467I44     [nan, 93.0]
4FMMO6     [97.0, nan]
4INBRV     [96.0, nan]
4QPNEA     [nan, 97.0]
4USS5P     [nan, 95.0]
5AHHQ9     [nan, 97.0]
5EM4RO     [93.0, nan]
5IE4NI     [91.0, nan]
5XBR59     [nan, 95.0]
6879G2     [nan, 84.0]
6A44G3     [93.0, nan]
6F1PTH     [98.0, nan]
6I3HKA     [nan, 91.0]
6JJBGL     [nan, 92.0]
7INVX5     [91.0, nan]
8LDTFT     [nan, 97.0]
8R9NH7    [100.0, nan]
91BJ5Q     [90.0, nan]
942H64     [91.0, nan]
9ODFA2     [93.0, nan]
9VMVYU     [98.0, nan]
AX1U3L     [nan, 92.0]
BQTIIK     [97.0, nan]
CV71CI     [nan, 97.0]
CZLOYU     [nan, 96.0]
D5QJWU     [97.0, nan]
DBC41Q     [nan, 93.0]
DS0QOH     [76.0, nan]
DSHXS6     [nan, 96.0]
E06QGZ     [97.0, nan]
FLGZW4     [91.0, nan]


**<font color='red'>c. Investigate the duplicate student ID records with different final_test values (52 records)**</font> </br>

In [ ]:
# View the data of the duplicate student ID records with different final_test values

final_test_conflicts = df[
    df["student_id"].isin(
        comparison.index[comparison["final_test"] > 1]
    )
].sort_values("student_id")

final_test_conflicts[
    ["student_id", "final_test"]
].head(20)

,student_id,final_test
2763,11V0BQ,NaN
12347,11V0BQ,79.0
905,2KKLZW,NaN
2204,2KKLZW,68.0
5039,2L5I46,80.0
14398,2L5I46,NaN
4897,34XQIE,NaN
12762,34XQIE,68.0
6520,3788PY,NaN
12663,3788PY,70.0


In [ ]:
# View final_test values per student_id

final_test_summary = (
    final_test_conflicts
    .groupby("student_id")["final_test"]
    .agg(list)
)

final_test_summary.head(52)

student_id
11V0BQ    [nan, 79.0]
2KKLZW    [nan, 68.0]
2L5I46    [80.0, nan]
34XQIE    [nan, 68.0]
3788PY    [nan, 70.0]
3F0S0J    [nan, 55.0]
3V7O8I    [nan, 93.0]
49LXTX    [57.0, nan]
56L4FT    [44.0, nan]
65XOE2    [65.0, nan]
780QOV    [54.0, nan]
7IZB0S    [79.0, nan]
7XH8AD    [55.0, nan]
86KRE2    [nan, 84.0]
9GIGZA    [69.0, nan]
9HW69M    [66.0, nan]
AWF13G    [nan, 87.0]
C8LUW0    [nan, 70.0]
CDCHI6    [60.0, nan]
DO7UAH    [48.0, nan]
E2SD8H    [74.0, nan]
EP44QF    [98.0, nan]
GD80JW    [nan, 71.0]
GYH9Y9    [52.0, nan]
H7N3F1    [54.0, nan]
H9KJUR    [nan, 56.0]
HL7EEZ    [68.0, nan]
J759E4    [61.0, nan]
JGWFQ8    [nan, 93.0]
KC8YO4    [nan, 80.0]
KF8BR9    [nan, 48.0]
KNWL5Z    [nan, 73.0]
L410MC    [49.0, nan]
MP3IHL    [57.0, nan]
NAL739    [54.0, nan]
NWKTGU    [79.0, nan]
O8QXHI    [62.0, nan]
PNJV5Q    [74.0, nan]
SGCHM0    [47.0, nan]
TE1RAN    [nan, 75.0]
U9SOUN    [64.0, nan]
UCYK4W    [55.0, nan]
UOGCK0    [nan, 79.0]
VG72MH    [nan, 78.0]
VWDRAH    [nan, 76.0]

##### <font color='blue'>**KEY OBSERVATIONS for attendance_rate & final_test:**</font><br>

1. The pattern is consistently: [NaN, valid value] or [valid value, NaN]
2. Data shows that the differences in values are not conflicting measurements. Instead, they're duplicated student records with complementary missingness.
3. Clean-up Strategy: To consolidate the 900 duplicate students into one record per student_id, preserving the non-null value wherever one duplicate contains the missing information.


**<font color='red'>d. Clean-up duplicate records for attendance_rate & final_test.**</font> </br>

In [76]:
# Create a new dataframe
df_clean = df.copy()

In [77]:
# Populate attendance_rate with the non-null value of the duplicate record
df_clean["attendance_rate"] = (
    df_clean.groupby("student_id")["attendance_rate"]
    .transform(lambda x: x.fillna(x.dropna().iloc[0]) if x.notna().any() else x)
)

df_clean

,index,number_of_siblings,direct_admission,CCA,learning_style,student_id,gender,tuition,final_test,n_male,n_female,age,hours_per_week,attendance_rate,sleep_time,wake_time,mode_of_transport,bag_color
0,0,0,Yes,Sports,Visual,ACN2BE,Female,No,69.0,14.0,2.0,16.0,10.0,91.0,22:00,6:00,private transport,yellow
1,1,2,No,Sports,Auditory,FGXIIZ,Female,No,47.0,4.0,19.0,16.0,7.0,94.0,22:30,6:30,private transport,green
2,2,0,Yes,None,Visual,B9AI9F,Male,No,85.0,14.0,2.0,15.0,8.0,92.0,22:30,6:30,private transport,white
3,3,1,No,Clubs,Auditory,FEVM1T,Female,Yes,64.0,2.0,20.0,15.0,18.0,NaN,21:00,5:00,public transport,yellow
4,4,0,No,Sports,Auditory,AXZN2E,Male,No,66.0,24.0,3.0,16.0,7.0,95.0,21:30,5:30,public transport,yellow
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
15895,15895,1,No,Clubs,Visual,XPECN2,Female,No,56.0,12.0,14.0,16.0,9.0,96.0,22:00,6:00,private transport,black
15896,15896,1,Yes,None,Auditory,7AMC7S,Male,Yes,85.0,17.0,5.0,16.0,7.0,91.0,22:30,6:30,private transport,white
15897,15897,1,Yes,Sports,Auditory,XKZ6VN,Female,Yes,76.0,7.0,10.0,15.0,7.0,93.0,23:00,7:00,walk,red
15898,15898,1,No,Clubs,Visual,2OU4UQ,Male,Yes,45.0,18.0,12.0,16.0,3.0,94.0,23:00,7:00,walk,yellow


In [78]:
# Populate final_test with the non-null value of the duplicate record
df_clean["final_test"] = (
    df_clean.groupby("student_id")["final_test"]
    .transform(lambda x: x.fillna(x.dropna().iloc[0]) if x.notna().any() else x)
)

df_clean

,index,number_of_siblings,direct_admission,CCA,learning_style,student_id,gender,tuition,final_test,n_male,n_female,age,hours_per_week,attendance_rate,sleep_time,wake_time,mode_of_transport,bag_color
0,0,0,Yes,Sports,Visual,ACN2BE,Female,No,69.0,14.0,2.0,16.0,10.0,91.0,22:00,6:00,private transport,yellow
1,1,2,No,Sports,Auditory,FGXIIZ,Female,No,47.0,4.0,19.0,16.0,7.0,94.0,22:30,6:30,private transport,green
2,2,0,Yes,None,Visual,B9AI9F,Male,No,85.0,14.0,2.0,15.0,8.0,92.0,22:30,6:30,private transport,white
3,3,1,No,Clubs,Auditory,FEVM1T,Female,Yes,64.0,2.0,20.0,15.0,18.0,NaN,21:00,5:00,public transport,yellow
4,4,0,No,Sports,Auditory,AXZN2E,Male,No,66.0,24.0,3.0,16.0,7.0,95.0,21:30,5:30,public transport,yellow
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
15895,15895,1,No,Clubs,Visual,XPECN2,Female,No,56.0,12.0,14.0,16.0,9.0,96.0,22:00,6:00,private transport,black
15896,15896,1,Yes,None,Auditory,7AMC7S,Male,Yes,85.0,17.0,5.0,16.0,7.0,91.0,22:30,6:30,private transport,white
15897,15897,1,Yes,Sports,Auditory,XKZ6VN,Female,Yes,76.0,7.0,10.0,15.0,7.0,93.0,23:00,7:00,walk,red
15898,15898,1,No,Clubs,Visual,2OU4UQ,Male,Yes,45.0,18.0,12.0,16.0,3.0,94.0,23:00,7:00,walk,yellow


In [79]:
#Remove duplicate student records

df_clean = df_clean.drop_duplicates(
    subset="student_id",
    keep="first"
).copy()

print("Rows after consolidation:", len(df_clean))
print("Unique students:", df_clean["student_id"].nunique())

Rows after consolidation: 15000
Unique students: 15000


In [80]:
print("Duplicate student IDs:",
      df_clean["student_id"].duplicated().sum())

print("\nMissing values:")
print(
    df_clean[
        ["attendance_rate", "final_test"]
    ].isnull().sum()
)

Duplicate student IDs: 0

Missing values:
attendance_rate    688
final_test         441
dtype: int64


**<font color='red'>e. Check the difference of 54 records reduction of missing final_test records.**</font> </br>

In [81]:
# Students with missing final_test in original data
original_missing_final = set(
    df.loc[df["final_test"].isna(), "student_id"]
)

# Students still missing final_test after consolidation
remaining_missing_final = set(
    df_clean.loc[df_clean["final_test"].isna(), "student_id"]
)

recovered_final = original_missing_final - remaining_missing_final

print("Students whose final_test was recovered:", len(recovered_final))

Students whose final_test was recovered: 52


In [82]:
# Get only duplicated student records
dup_students = df[
    df["student_id"].duplicated(keep=False)
]

# Find duplicate students where BOTH final_test values are missing
both_final_missing = (
    dup_students
    .groupby("student_id")
    .filter(lambda x: x["final_test"].isna().all())
)

both_final_missing[
    ["student_id", "final_test"]
].sort_values("student_id")

,student_id,final_test
2348,BLHFEH,NaN
6275,BLHFEH,NaN
1798,T9PF1C,NaN
11405,T9PF1C,NaN


In [83]:
print(
    "Students with final_test missing in both duplicate records:",
    both_final_missing["student_id"].nunique()
)

Students with final_test missing in both duplicate records: 2


##### <font color='blue'>**RECONCILIATION of final_test:**</font><br>

Original missing final_test rows:               495  <br>
Less: scores genuinely recovered:                52   <br>
Less: redundant NaN rows removed:                 2   <br>
---------------------------------------------------  <br>
Remaining students with missing final_test:      441  <br>


##### <font color='blue'>**UPDATED NUMBER of MISSING VALUES**</font><br>
1. final_test: Original missing records of 495 (3.11%) is now down to 441 (2.94%)
2. attendance_rate: Original missing records of 778 (4.9%) is now down to 688 (4.6%)


In [91]:
# Create a new dataframe for the dedup dataset

df_dedup = df_clean.copy()

print(f"Number of rows: {df_dedup.shape[0]}")
print(f"Number of columns: {df_dedup.shape[1]}")

print("Number of unique student ids:",
    df_dedup["student_id"].nunique())

Number of rows: 15000
Number of columns: 18
Number of unique student ids: 15000


**<font color='red'>2. INVALID STUDENT AGE**</font> </br>

**<font color='red'>a. Before the clean-up and dedeup process on the duplicated student_id.**</font> </br>

In [ ]:
# Count the number of records per value
df["age"].value_counts().sort_index()

age
-5.0        4
-4.0        1
 5.0      216
 6.0      230
 15.0    7726
 16.0    7723
Name: count, dtype: int64

In [86]:
# View the records of the invalid ages 
invalid_age = df[
    (df["age"] < 10) |
    (df["age"] > 20)
]

print("Invalid/suspicious age records:", len(invalid_age))

invalid_age[[
    "student_id",
    "age",
    "final_test",
    "hours_per_week",
    "attendance_rate"
]].head(10)

Invalid/suspicious age records: 451


,student_id,age,final_test,hours_per_week,attendance_rate
30,7882P8,6.0,NaN,14.0,91.0
60,1B3QSS,5.0,40.0,5.0,98.0
61,D8YD78,6.0,57.0,11.0,95.0
81,ANUEDH,5.0,57.0,6.0,96.0
82,7EPPQG,5.0,70.0,10.0,96.0
83,EAZDA8,6.0,52.0,8.0,95.0
114,2QYBOU,6.0,72.0,10.0,97.0
140,TDLKDV,5.0,72.0,10.0,99.0
142,2RE5VB,5.0,61.0,7.0,91.0
145,FR8U8M,6.0,41.0,8.0,96.0


**<font color='red'>b. After the clean-up and dedeup process on the duplicated student_id.**</font> </br>

In [92]:
age_analysis = (
    df_dedup
    .groupby("age")
    .agg(
        students=("student_id", "count"),
        mean_score=("final_test", "mean"),
        median_score=("final_test", "median"),
        mean_study_hours=("hours_per_week", "mean"),
        mean_attendance=("attendance_rate", "mean")
    )
    .round(2)
)

age_analysis

,students,mean_score,median_score,mean_study_hours,mean_attendance
age,,,,,
-5.0,4,69.50,70.5,12.25,91.75
-4.0,1,48.00,48.0,2.00,94.00
5.0,209,66.51,67.0,10.09,92.52
6.0,221,66.01,67.0,10.68,93.33
15.0,7273,67.16,68.0,10.40,93.42
16.0,7292,67.27,68.0,10.23,93.17


**<font color='red'>c. Convert all five impossible values/categories of age to NaN.**</font> </br>

In [ ]:
df_clean.loc[
    ~df_clean["age"].isin([15, 16]),
    "age"
] = np.nan

In [96]:
df_clean["age"].value_counts(dropna=False)

age
16.0    7292
15.0    7273
NaN      435
Name: count, dtype: int64

##### <font color='blue'>**KEY OBSERVATIONS for age**</font><br>
1. After the clean-up and de-dup of student ids:  <br>
- 435 students had implausible ages of -5, -4, 5, or 6 (2.9%) <br>
- 14,565 students were aged 15 or 16 (97.1%) <br>
2. Students recorded as ages 5 and 6 exhibited academic characteristics broadly similar to the valid-age population, suggesting these values are data-quality errors rather than genuine ages. Although the values may represent truncated ages (e.g., 5 → 15), the original age cannot be established with certainty. <br>
3. Therefore, implausible ages were treated as unknown (NaN) rather than manually corrected. <br>

##### **Step 5. Standardise CCA and tuition values**

##### A. CCA

In [ ]:
# Change to title case
df_clean["CCA"] = df_clean["CCA"].str.title()

In [ ]:
# View the unique values and counts
df_clean["CCA"].value_counts()

CCA
Clubs     3805
Sports    3760
None      3730
Arts      3705
Name: count, dtype: int64

##### B. tuition

In [ ]:
# Replace Y to Yes and N to No
df_clean["tuition"] = df_clean["tuition"].replace({
    "Y": "Yes",
    "N": "No"
})

In [ ]:
# View the unique values and counts
df_clean["tuition"].value_counts()

tuition
Yes    8501
No     6499
Name: count, dtype: int64

In [ ]:
# Confirm that standardisation done hasn't changed the row count of the dataset
print("Total Rows:", len(df_clean))
print("Unique students:", df_clean["student_id"].nunique())

Total Rows: 15000
Unique students: 15000


##### <font color='blue'>**CATEGORICAL STANDARDIZATION**</font><br>

1. Inconsistent category labels were identified in CCA (Clubs/CLUBS, Sports/SPORTS, etc.) and tuition (Yes/Y, No/N). 
2. These were standardised into four consistent CCA categories and two tuition categories respectively. 
3. The transformation did not affect the dataset size, which remained at 15,000 unique students.

##### **Step 6. Feature engineering for sleep_duration**

In [ ]:
# Convert sleep and wake times to datetime
sleep_dt = pd.to_datetime(
    df_clean["sleep_time"],
    format="%H:%M"
)

wake_dt = pd.to_datetime(
    df_clean["wake_time"],
    format="%H:%M"
)

# Calculate duration in hours, accounting for crossing midnight
df_clean["sleep_duration"] = (
    (wake_dt - sleep_dt).dt.total_seconds() / 3600
) % 24

df_clean.head(10)

,index,number_of_siblings,direct_admission,CCA,learning_style,student_id,gender,tuition,final_test,n_male,n_female,age,hours_per_week,attendance_rate,sleep_time,wake_time,mode_of_transport,bag_color,sleep_duration
0,0,0,Yes,Sports,Visual,ACN2BE,Female,No,69.0,14.0,2.0,16.0,10.0,91.0,22:00,6:00,private transport,yellow,8.0
1,1,2,No,Sports,Auditory,FGXIIZ,Female,No,47.0,4.0,19.0,16.0,7.0,94.0,22:30,6:30,private transport,green,8.0
2,2,0,Yes,None,Visual,B9AI9F,Male,No,85.0,14.0,2.0,15.0,8.0,92.0,22:30,6:30,private transport,white,8.0
3,3,1,No,Clubs,Auditory,FEVM1T,Female,Yes,64.0,2.0,20.0,15.0,18.0,NaN,21:00,5:00,public transport,yellow,8.0
4,4,0,No,Sports,Auditory,AXZN2E,Male,No,66.0,24.0,3.0,16.0,7.0,95.0,21:30,5:30,public transport,yellow,8.0
5,5,0,No,Arts,Visual,BA6R14,Female,No,57.0,9.0,12.0,15.0,11.0,96.0,22:30,6:30,private transport,red,8.0
6,6,2,Yes,None,Visual,D5WGTI,Male,No,69.0,12.0,3.0,16.0,15.0,93.0,21:30,5:30,public transport,green,8.0
7,7,0,No,Sports,Visual,HTP8CW,Male,No,76.0,20.0,2.0,15.0,3.0,97.0,21:00,5:00,public transport,green,8.0
8,8,0,No,Arts,Auditory,U3YRTC,Male,No,57.0,20.0,7.0,15.0,15.0,98.0,22:00,6:00,private transport,red,8.0
9,9,2,No,Arts,Auditory,3MOMA6,Male,Yes,60.0,13.0,9.0,16.0,16.0,NaN,22:30,6:30,private transport,green,8.0


In [ ]:
# View summary statistics
df_clean["sleep_duration"].describe()

count    15000.000000
mean         7.847600
std          0.599776
min          4.000000
25%          8.000000
50%          8.000000
75%          8.000000
max          8.000000
Name: sleep_duration, dtype: float64

In [ ]:
# View the unique combinations of sleep_time, wake_time and sleep_duration
sleep_combinations = (
    df_clean
    .groupby(
        ["sleep_time", "wake_time", "sleep_duration"]
    )
    .size()
    .reset_index(name="student_count")
    .sort_values(
        ["sleep_duration", "sleep_time", "wake_time"]
    )
)

sleep_combinations

,sleep_time,wake_time,sleep_duration,student_count
5,1:00,5:00,4.0,37
8,1:30,5:30,4.0,27
21,2:00,6:00,4.0,31
23,2:30,6:30,4.0,24
24,3:00,7:00,4.0,22
0,0:00,5:00,5.0,62
3,0:30,5:30,5.0,37
6,1:00,6:00,5.0,37
9,1:30,6:30,5.0,40
22,2:00,7:00,5.0,44


In [ ]:
# Count the number of unique combinations
print(
    "Number of unique sleep/wake combinations:",
    len(sleep_combinations)
)

Number of unique sleep/wake combinations: 25


In [ ]:
# Count the number of records per value
df_clean["sleep_duration"].value_counts().sort_index()

sleep_duration
4.0      141
5.0      220
6.0      223
7.0      616
8.0    13800
Name: count, dtype: int64

In [ ]:
# View sample records (least sleep_duration)
df_clean[
    ["sleep_time", "wake_time", "sleep_duration"]
].sort_values("sleep_duration").head(10)

,sleep_time,wake_time,sleep_duration
8023,1:30,5:30,4.0
1754,1:30,5:30,4.0
9488,2:30,6:30,4.0
9510,2:30,6:30,4.0
12177,2:30,6:30,4.0
15042,2:00,6:00,4.0
12184,2:00,6:00,4.0
7286,2:00,6:00,4.0
11074,1:00,5:00,4.0
3615,2:00,6:00,4.0


In [ ]:
# View sample records (most sleep_duration)
df_clean[
    ["sleep_time", "wake_time", "sleep_duration"]
].sort_values("sleep_duration", ascending=False).head(10)

,sleep_time,wake_time,sleep_duration
0,22:00,6:00,8.0
10238,21:30,5:30,8.0
10146,22:00,6:00,8.0
10147,22:30,6:30,8.0
10148,21:30,5:30,8.0
10149,22:00,6:00,8.0
10150,23:00,7:00,8.0
10151,22:30,6:30,8.0
10152,21:30,5:30,8.0
10154,21:30,5:30,8.0


In [ ]:
# Check the relationship between sleep_duration and final_test
sleep_analysis = (
    df_clean
    .groupby("sleep_duration")
    .agg(
        students=("student_id", "count"),
        mean_score=("final_test", "mean"),
        median_score=("final_test", "median"),
        std_score=("final_test", "std")
    )
    .round(2)
)

sleep_analysis

,students,mean_score,median_score,std_score
sleep_duration,,,,
4.0,141,43.83,44.0,3.54
5.0,220,45.07,46.0,3.92
6.0,223,45.64,47.0,3.52
7.0,616,61.58,60.0,15.69
8.0,13800,68.38,69.0,13.31


In [113]:
df_clean.groupby("sleep_duration")["final_test"].describe().round(2)

,count,mean,std,min,25%,50%,75%,max
sleep_duration,,,,,,,,
4.0,136.0,43.83,3.54,32.0,42.0,44.0,46.0,50.0
5.0,214.0,45.07,3.92,32.0,43.0,46.0,48.0,50.0
6.0,218.0,45.64,3.52,32.0,43.0,47.0,48.0,50.0
7.0,598.0,61.58,15.69,32.0,48.0,60.0,74.0,99.0
8.0,13393.0,68.38,13.31,32.0,58.0,69.0,78.0,100.0


In [114]:
for hours in [4, 5, 6, 7, 8]:
    print(f"\n--- {hours} hours ---")
    print(
        df_clean.loc[
            df_clean["sleep_duration"] == hours,
            "final_test"
        ].value_counts().sort_index()
    )


--- 4 hours ---
final_test
32.0     1
35.0     1
36.0     2
38.0     3
39.0    10
40.0    12
41.0     4
42.0     8
43.0    16
44.0    18
45.0    26
46.0     4
47.0     7
48.0     8
49.0     9
50.0     7
Name: count, dtype: int64

--- 5 hours ---
final_test
32.0     2
34.0     2
35.0     1
36.0     1
37.0     4
38.0     5
39.0     7
40.0     9
41.0     8
42.0    10
43.0    13
44.0    16
45.0    25
46.0    17
47.0    24
48.0    24
49.0    26
50.0    20
Name: count, dtype: int64

--- 6 hours ---
final_test
32.0     1
34.0     1
36.0     1
37.0     2
38.0     3
39.0     4
40.0     5
41.0    15
42.0    12
43.0    13
44.0    17
45.0    18
46.0    13
47.0    30
48.0    29
49.0    33
50.0    21
Name: count, dtype: int64

--- 7 hours ---
final_test
32.0    2
34.0    1
36.0    3
37.0    2
38.0    4
       ..
95.0    2
96.0    3
97.0    1
98.0    1
99.0    1
Name: count, Length: 65, dtype: int64

--- 8 hours ---
final_test
32.0      4
34.0      6
35.0      2
36.0      7
37.0      8
         ..
9

##### <font color='blue'>**KEY OBSERVATIONS for sleep_duration**</font><br>
1. Students sleeping 8 hours average about 24.6 points higher than those sleeping 4 hours. <br>
2. We cannot deduce that “more sleep causes better math scores”. The analysis only establishes a strong association, not causation, and there may be other variables driving part of the pattern. <br>
3. There appears to be a substantial divide between ≤6 hours and ≥7 hours. <br>
4. Each sleep_duration has exactly five possible sleep/wake combinations.
5. For the 8 hours, those five combinations are evenly distributed: 2,706–2,810 students each.

### <font color = green> **PART B: EXPLORATORY DATA ANALYSIS**</br></font>

In [ ]:
df_clean["final_test"].describe()